src/ucs/memory/rcache.c

```diff
@@ -1155,6 +1221,14 @@ static UCS_CLASS_CLEANUP_FUNC(ucs_rcache_t)
 {
     ucs_vfs_obj_remove(self);
     ucs_rcache_global_list_remove(self);
+    pthread_mutex_lock(&ucs_rcache_global_context.lock);
+    if (ucs_list_is_empty(&ucs_rcache_global_context.list)) {
```

Comment:
only checking if the list is empty should be inside the lock
BTW, maybe add helper function "ucs_rcache_global_list_remove"